src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp

```diff
@@ -60,39 +82,62 @@ namespace Sources {
 struct VortexPerturbation {
-  /// The \f$z-\f$component of the vortex flow velocity.
-  struct VelocityAlongZ {
-    using type = Scalar<DataVector>;
-    static std::string name() noexcept { return "VelocityAlongZ"; }
-  };
-
-  /// The derivative w.r.t. \f$z\f$ of the \f$z-\f$component of the velocity.
-  struct DzVelocityAlongZ {
-    using type = Scalar<DataVector>;
-    static std::string name() noexcept { return "DzVelocityAlongZ"; }
-  };
-
-  using argument_tags =
-      tmpl::list<Tags::MassDensity<DataVector>,
-                 Tags::MomentumDensity<DataVector, 3, Frame::Inertial>,
-                 Tags::EnergyDensity<DataVector>, Tags::Pressure<DataVector>,
-                 VelocityAlongZ, DzVelocityAlongZ>;
-
-  static void apply(
-      gsl::not_null<Scalar<DataVector>*> mass_density_source,
-      gsl::not_null<tnsr::I<DataVector, 3>*> momentum_density_source,
-      gsl::not_null<Scalar<DataVector>*> energy_density_source,
-      const Scalar<DataVector>& vortex_mass_density,
-      const tnsr::I<DataVector, 3>& vortex_momentum_density,
-      const Scalar<DataVector>& vortex_energy_density,
-      const Scalar<DataVector>& vortex_pressure,
-      const Scalar<DataVector>& vortex_velocity_z,
-      const Scalar<DataVector>& dz_vortex_velocity_z) noexcept;
+  VortexPerturbation() noexcept = default;
+  VortexPerturbation(const VortexPerturbation& /*rhs*/) = default;
+  VortexPerturbation& operator=(const VortexPerturbation& /*rhs*/) = default;
+  VortexPerturbation(VortexPerturbation&& /*rhs*/) noexcept = default;
+  VortexPerturbation& operator=(VortexPerturbation&& /*rhs*/) noexcept =
+      default;
+  ~VortexPerturbation() = default;
+
+  VortexPerturbation(double adiabatic_index, double perturbation_amplitude,
+                     const std::array<double, Dim>& vortex_center,
+                     const std::array<double, Dim>& vortex_mean_velocity,
+                     double vortex_strength) noexcept;
+
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& /*p*/) noexcept;  // NOLINT
+
+  using sourced_variables = typename detail::get_sourced_variables<Dim>::type;
+
+  using argument_tags = typename detail::get_argument_tags<Dim>::type;
+
+  // Overload required for 2-D simulations, where no variable is sourced.
+  void apply() const noexcept;
+
+  // Function to be used in 3-D.
+  void apply(gsl::not_null<Scalar<DataVector>*> source_mass_density_cons,
+             gsl::not_null<tnsr::I<DataVector, Dim>*> source_momentum_density,
+             gsl::not_null<Scalar<DataVector>*> source_energy_density,
+             const tnsr::I<DataVector, Dim>& x, double time) const noexcept;
+
+ private:
+  template <size_t SpatialDim>
+  friend bool
+  operator==(  // NOLINT (clang-tidy: readability-redundant-declaration)
+      const VortexPerturbation<SpatialDim>& lhs,
+      const VortexPerturbation<SpatialDim>& rhs) noexcept;
+
+  double adiabatic_index_ = std::numeric_limits<double>::signaling_NaN();
```

Comment:
Instead of creating the solution every time `apply` is called, just pass it to `apply` and then you don't need these member variables